### PR TITLE
update weightstreetnet docs to fix #245

### LINKED ISSUE #245: Improve documentation of weight_streetnet
Hi @mpadge! I’ve created this issue since I would like to ask for better documentation regarding the behaviour of `weight_streetnet`. For example:

``` r
# packages
library(dodgr)
library(sf)
#> Linking to GEOS 3.11.2, GDAL 3.7.2, PROJ 9.3.0; sf_use_s2() is TRUE

# Just one segment
toy <- st_as_sf(
  data.frame(osm_id = 1, highway = "cycleway"), 
  geometry = st_sfc(
    st_linestring(rbind(c(0, 0), c(1, 1)))
  )
)
weight_streetnet(toy)[1:8]
#>   geom_num edge_id from_id from_lon from_lat to_id to_lon to_lat
#> 1        1       1       0        0        0     1      1      1
#> 2        1       2       1        1        1     0      0      0
```

Everything is fine! However, if I add a `oneway` field and set its value equal to “yes”, the function returns just one edge.

``` r
toy <- st_as_sf(
  data.frame(osm_id = 1, highway = "cycleway", oneway = "yes"), 
  geometry = st_sfc(
    st_linestring(rbind(c(0, 0), c(1, 1)))
  )
)
weight_streetnet(toy)[1:8]
#>   geom_num edge_id from_id from_lon from_lat to_id to_lon to_lat
#> 1        1       1       0        0        0     1      1      1
```

I totally understand (and I like!) this behaviour but, unfortunately, I think it’s a bit surprising. For example, I cannot find anywhere in the docs:
1) which field names modify the behaviour of `weight_streetnet`?
2) what are the possible values for these fields?

I think this information is crucial for users. Apologies for the noise if this is already documented somewhere, but I could find any reference. 

Thanks!

<sup>Created on 2024-09-20 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by mpadge**

### CODE REVIEW

File: R/weight-streetnet.R [hunk 1]
```diff
@@ -1,9 +1,64 @@
...
+#' street network according to a named profile, selected from (foot, horse,
+#' wheelchair, bicycle, moped, motorcycle, motorcar, goods, hgv, psv), or a
+#' cusstomized version dervied from those.
+#'
+#' @details The structure of networks generated by this function is dependent
+#' on many aspects of the input network, and in particular on specific
+#' key-value pairs defined in the underlying OpenStreetMap (OSM) data.
+#'
+#' OSM keys which can be used to determine restrictions, and so may be added to
+#' the `keep_cols` argument, include:
```
Comment (agila5) [inline R/weight-streetnet.R:13]: Currently, it looks like that the behaviour of `weight_streetnet()` depends on the following fields (access, bicycle, foot, ...) only when they are specified in the `keep_cols` argument, but this is not the case. Right? For example, the reprex included in #251 shows that the function creates one-way edges even when `keep_cols = NULL`.

File: R/weight-streetnet.R [hunk 2]
```diff
@@ -1,9 +1,64 @@
...
+#' \item "oneway"
+#' \item "restriction"
+#' \item "toll"
+#' \item "vehicle"
+#' }
+#'
+#' Realistic routing including factors such as access restrictions, turn
+#' penalties, and effects of incline, can only be implemented when the objects
+#' passed to `weight_streetnet` are of \pkg{sc} ("silicate") format, generated
+#' with \link{dodgr_streetnet_sc}. Restrictions applied to ways (in OSMM
```
Comment (agila5) [inline R/weight-streetnet.R:30]: Typo, it should be OSM instead of OSMM

File: R/weight-streetnet.R [hunk 3]
```diff
@@ -1,9 +1,64 @@
 #' Weight a street network according to a specified weighting profile.
 #'
-#' Weight (or re-weight) an \pkg{sf} or `SC` (`silicate`)-formatted OSM street
-#' network according to a named profile, selected from (foot, horse, wheelchair,
-#' bicycle, moped, motorcycle, motorcar, goods, hgv, psv), or a customized
-#' version derived from those.
+#' Weight (or re-weight) an \pkg{sf} or \pkg{silicate} *("SC") formatted OSM
+#' street network according to a named profile, selected from (foot, horse,
+#' wheelchair, bicycle, moped, motorcycle, motorcar, goods, hgv, psv), or a
+#' cusstomized version dervied from those.
```
Comment (agila5) [inline R/weight-streetnet.R:6]: This paragraph still has the same typos that were fixed by #248

File: R/weight-streetnet.R [hunk 4]
```diff
@@ -1,9 +1,64 @@
...
+#' \item "toll"
+#' \item "vehicle"
+#' }
+#'
+#' Realistic routing including factors such as access restrictions, turn
+#' penalties, and effects of incline, can only be implemented when the objects
+#' passed to `weight_streetnet` are of \pkg{sc} ("silicate") format, generated
+#' with \link{dodgr_streetnet_sc}. Restrictions applied to ways (in OSMM
+#' terminology) may be controlled by ensuring specific columns are retained in
+#' the `dodgr` network with the `keep_cols` argument. For example, restrictions
```
Comment (agila5) [inline R/weight-streetnet.R:32]: This behaviour holds only with `sc` objects, right? If so, I would document which fields influence only `sf` objects (I guess none), which fields influence only `sc` objects, which fields influence both.

File: R/weight-streetnet.R [hunk 5]
```diff
@@ -1,9 +1,64 @@
...
+#' they have been saved and loaded with the \link{dodgr_save_streetnet} and
+#' \link{dodgr_load_streetnet} functions. (This is because the weighted
+#' streetnets also have accompanying data stored in a local temporary cache
+#' directory; attempting to pass a weighted street network without these
+#' accompanying cache files will generally error.)
+#'
+#' Some key-value pairs may also directly influence not just the value of the
+#' graph produced by this function, but also its size. Among these are "oneway"
+#' flags. Without these flags, each edge will be represented in *directed* form,
+#' and so as two rows of the graph for A -> B and B -> A. If a way is tagged in
```
Comment (agila5) [inline R/weight-streetnet.R:54]: "for" or "from"?

File: R/weight-streetnet.R [hunk 6]
```diff
@@ -1,9 +1,64 @@
...
+#' OSM as "oneway" = "yes", and if oneway flags are respected for a chosen
+#' weighting profile (which, for example, they are generally not for pedestrian
+#' or "foot" weighting), then only one edge will be returned representing
+#' travel in the direction permitted within the OSM data. Thus weighting a
+#' network which includes "oneway" flags, and using a weighting profile which
+#' respects these, will generate a graph with fewer rows than a graph produced
+#' by ignoring those "oneway" flags.
```
Comment (agila5) [inline R/weight-streetnet.R:61]: LGTM!

File: R/weight-streetnet.R (hunk 2 above)
Comment (mpadge) [reply R/weight-streetnet.R]: fixed, thanks

File: R/weight-streetnet.R (hunk 3 above)
Comment (mpadge) [reply R/weight-streetnet.R]: Fixed; thanks!

File: R/weight-streetnet.R (hunk 5 above)
Comment (mpadge) [reply R/weight-streetnet.R:54]: "for" but obviously need to explain better, which I've now done

File: R/weight-streetnet.R (hunk 4 above)
Comment (mpadge) [reply R/weight-streetnet.R]: The whole list of keys has now been completely rewritten again. The biggest difference between sf and sc is in turn restrictions and effects of elevation, both of which are only possible in sc. This is now explicitly explained. Both of those work without specifying any additional `keep_cols`. I'll resolve this comment now, and ask for your review on this PR again, thanks

**Comment by mpadge** [comment]
@agila5 Thanks for your excellent feedback. I've now largely rewritten the whole thing again. Would you please be kind enough to review it once again? Thanks!

**Comment by agila5** [comment]
Hi @mpadge and sorry for the delay. I've been very busy the last days. I'll review it right now!

File: R/weight-streetnet.R [hunk 7]
```diff
@@ -1,9 +1,99 @@
...
+#' street network according to a named profile, selected from (foot, horse,
+#' wheelchair, bicycle, moped, motorcycle, motorcar, goods, hgv, psv), or a
+#' customized version derived from those.
+#'
+#' @details The structure of networks generated by this function is dependent
+#' on many aspects of the input network, and in particular on specific
+#' key-value pairs defined in the underlying OpenStreetMap (OSM) data.
+#'
+#' Many key-value pairs influence the resultant network through being used in
+#' specified weighting profiles. Keys used in weighting profiles are always
```
Comment (agila5) [inline R/weight-streetnet.R:13]: Just one very minor thing I don't understand. This part of text means that if OSM data includes a column named `bridleway`, `cycleway`, ... and that weighting profile is included in the corresponding mode of transport, that column is always retained in the output object, right? 

I don't understand the following then: 

``` r
library(dodgr)
library(sf)
#> Linking to GEOS 3.11.2, GDAL 3.7.2, PROJ 9.3.0; sf_use_s2() is TRUE

# The default mode of transport is "bicycle" which contains the following
# highway types:
weighting_profiles$weighting_profiles[67:88, "way"]
#>  [1] "motorway"       "trunk"          "primary"        "secondary"     
#>  [5] "tertiary"       "unclassified"   "residential"    "service"       
#>  [9] "track"          "cycleway"       "path"           "steps"         
#> [13] "ferry"          "living_street"  "bridleway"      "footway"       
#> [17] "pedestrian"     "motorway_link"  "trunk_link"     "primary_link"  
#> [21] "secondary_link" "tertiary_link"

# For example
toy <- st_as_sf(
  data.frame(highway = "primary", cycleway = c("A", "B"), osm_id = 1:2), 
  geometry = st_sfc(
    st_linestring(rbind(c(0, 0), c(1, 1))), 
    st_linestring(rbind(c(1, 1), c(2, 2)))
  )
)
weight_streetnet(toy) # cycleway is retained
#>   geom_num edge_id from_id from_lon from_lat to_id to_lon to_lat        d
#> 1        1       1       0        0        0     1      1      1 156899.6
#> 2        1       2       1        1        1     0      0      0 156899.6
#> 3        2       3       1        1        1     2      2      2 156876.1
#> 4        2       4       2        2        2     1      1      1 156876.1
#>   d_weighted highway way_id component cycleway     time time_weighted
#> 1   224142.2 primary      1         1        A 37655.90      53794.14
#> 2   224142.2 primary      1         1        A 37655.90      53794.14
#> 3   224108.8 primary      2         1        B 37650.28      53786.11
#> 4   224108.8 primary      2         1        B 37650.28      53786.11

# Same structure but considering 
toy <- st_as_sf(
  data.frame(highway = "primary", path = c("A", "B"), osm_id = 1:2), 
  geometry = st_sfc(
    st_linestring(rbind(c(0, 0), c(1, 1))), 
    st_linestring(rbind(c(1, 1), c(2, 2)))
  )
)
weight_streetnet(toy) # path is not retained
#>   geom_num edge_id from_id from_lon from_lat to_id to_lon to_lat        d
#> 1        1       1       0        0        0     1      1      1 156899.6
#> 2        1       2       1        1        1     0      0      0 156899.6
#> 3        2       3       1        1        1     2      2      2 156876.1
#> 4        2       4       2        2        2     1      1      1 156876.1
#>   d_weighted highway way_id component     time time_weighted
#> 1   224142.2 primary      1         1 37655.90      53794.14
#> 2   224142.2 primary      1         1 37655.90      53794.14
#> 3   224108.8 primary      2         1 37650.28      53786.11
#> 4   224108.8 primary      2         1 37650.28      53786.11
```

<sup>Created on 2024-09-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Pull request merged**